modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/OnceLogger.java

```diff
@@ -0,0 +1,162 @@
+
+    /**
+     * Allow advanced users to disable the OnceLogger (more for performance tuning in hosted environments).
+     * This is really only useful or necessary if this implementation causes issues.
+     */
+    private static final String ENABLE_ONCE_LOGGER_PROPERTY = "org.openapitools.codegen.utils.oncelogger.enabled";
+
+    /**
+     * Allow advanced users to modify cache expiration of the OnceLogger (more for performance tuning in hosted environments)
+     */
+    private static final String EXPIRY_PROPERTY = "org.openapitools.codegen.utils.oncelogger.expiry";
+
+    /**
+     * The fully qualified class name of the <b>logger instance</b>,
+     * typically the logger class, logger bridge or a logger wrapper.
+     */
+    private static final String FQCN = OnceLogger.class.getName();
+
+    /**
+     * Gets the marker instance. This can be used by supported log implementations to filter/manage logs coming from
+     * this implementation differently than others (i.e. make them stand out since they're to be logged once).
+     */
+    private static final Marker MARKER = MarkerFactory.getMarker("ONCE");
+
+    /**
+     * The allowed size of the cache.
+     */
+    private static int maxCacheSize = Integer.parseInt(GlobalSettings.getProperty(CACHE_SIZE_PROPERTY, "200"));
+
+    /**
+     * The millis to expire a cached log message.
+     */
+    private static int expireMillis = Integer.parseInt(GlobalSettings.getProperty(EXPIRY_PROPERTY, "2000"));
+
+    /**
+     * The number of allowed repetitions.
+     */
+    private static int maxRepetitions = 1;
+
+    /**
+     * Internal message cache for logger decorated with the onceler.
+     */
+    private static Cache<String, Integer> messageCountCache;
+
+    OnceLogger(Logger logger) {
+        this(logger, FQCN);
+    }
+
+    OnceLogger(Logger logger, String fqcn) {
+        super(logger, fqcn);
+    }
+
+    static {
+        messageCountCache = Caffeine.newBuilder()
+                .maximumSize(maxCacheSize)
+                .expireAfterWrite(expireMillis, TimeUnit.MILLISECONDS)
+                .build();
+    }
+
+    public static Logger once(Logger logger) {
+        try {
+            if (Boolean.parseBoolean(GlobalSettings.getProperty(ENABLE_ONCE_LOGGER_PROPERTY, "true"))) {
+                return new OnceLogger(logger);
+            }
+        } catch (Exception ex) {
+            logger.warn("Unable to wrap logger instance in OnceLogger. Falling back to non-decorated implementation, which may be noisy.");
+        }
+        return logger;
+    }
+
+    /**
+     * Delegate to the appropriate method of the underlying logger.
+     *
+     * @param msg
+     */
+    @Override
+    public void trace(String msg) {
+        if (!isTraceEnabled() || !isTraceEnabled(MARKER)) return;
+
+        if (shouldLog(msg)) super.trace(MARKER, msg);
+    }
+
+    private boolean shouldLog(final String msg) {
```

Comment:
maybe a single operation if you want to avoid races?
`messageCountCache.asMap().merge(msg, 0, count -> count + 1)`

or an atomic counter if you want to avoid locking except on the initial mapping?

Reply:
Thanks for the quick and helpful review, Ben! I decided to go the `AtomicInteger` route because it's not critical functionality for this method and doesn't need repeated locking, I'm just trying to remove noise for users.

Reply:
I think that's best choice, too! 

However, it no longer writes into the map on every `shouldLog` call and the `expireAfterWrite` evicts based on the entry's creation time. The old behavior was to reset the expiration each time the method was called, so it requires no occurrences for the expiration duration. That could be done now using `expireAfterAccess` to reset on every read or write, which is fine as this is the sole usage.

I suspect you don't want to suppress an ongoing error for good once the threshold is reached, as that makes it appear to be over, and simply want to debounce over a time interval. This new expiration behavior might be more correct, but I'm unsure. You probably want to think a bit about exactly what expiration behavior you want and maybe write a quick test to verify that.

Reply:
You're correct that my intent was to debounce over a time internal. I think I mistakenly took `expireAfterWrite` to mean initial writes only, but looking at it after your comment I see that understanding wouldn't have made sense if I'd looked at the method a little more closely.

I'll definitely need to add a comment about intention on the cache member, and write a handful of tests. I'll have to think about usage for errors and other levels a little more. The way I wrote this, it'll require that someone explicitly decorates a Logger instance and I may just throw an error if it's used to log error messages, or just skip the cache evaluation for that method.